### PR TITLE
Enable accessing Kokkos::Array elements in constant expressions

### LINKED ISSUE #4905: Constexpr operator[] for Kokkos::Array?
Hi! I was wondering: why `Kokkos::Array`'s `operator[]` is not declared `constexpr`?
I'd like to be able to compute at compile time some coefficients and store them in a `Kokkos::Array`, and I end up having to write some ugly pieces of code in order to make this work (e.g. bypassing the API to access the C-array inside directly). I can't imagine a situation where having this operator available in a `constexpr` context could lead to trouble... Did I miss something?

### PR DESCRIPTION
**Pull request by dalg24**
Fix #4905 
The tricky part was the custom error messages for out of bounds accesses.
The string manipulation facility is a bit heavy handed but we will need something like that anyway to produce better error messages with memory access violations in `Kokkos::View`.
I decided to call abort from the host-side as well instead of throwing a runtime error.

### CODE REVIEW

File: core/src/impl/Kokkos_StringManipulation.hpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    }
+    return i;
+  }
+  return 0;
+}
+
+// compares two strings
+KOKKOS_FUNCTION constexpr int strcmp(const char *lhs, const char *rhs) {
+  if (lhs == nullptr || rhs == nullptr) {
+    return rhs != nullptr ? -1 : 1;
```
Comment (nliber) [inline core/src/impl/Kokkos_StringManipulation.hpp:76]: For `strcmp`, a null string is strictly less than an empty string.

Suggested change:

If both `lhs` and `rhs` have the same pointer value they should compare equal (return `0`), even if that value is the null pointer value.

One way to write that is:

```c++
if (lhs == rhs)
  return 0;
if (lhs == nullptr)
  return -1;
if (rhs == nullptr)
  return 1;
//... do the lexicographical comparison of elements
```

File: core/src/impl/Kokkos_StringManipulation.hpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    for (; *d != '\0'; ++d) {
+    }
+    while ((*d++ = *src++) != '\0') {
+    }
+  }
+  return dest;
+}
+
+// concatenates a certain amount of characters of two strings
+KOKKOS_FUNCTION constexpr char *strncat(char *dest, const char *src,
```
Comment (nliber) [inline core/src/impl/Kokkos_StringManipulation.hpp:145]: Same suggested change as for `strcat`.

File: core/unit_test/TestStringManipulation.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+//@HEADER
+*/
+
+#include <impl/Kokkos_StringManipulation.hpp>
+#include <climits>
+
+namespace {
+
+#define STATIC_ASSERT(cond) static_assert(cond, "")
+
```
Comment (nliber) [inline core/unit_test/TestStringManipulation.cpp:51]: We should also have string manipulation tests with null pointer values.

File: core/src/impl/Kokkos_StringManipulation.hpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    } while (--count != 0);
+    *d = '\0';
+  }
+  return dest;
+}
+//</editor-fold>
+
+//<editor-fold desc="Character conversions">
+template <class Unsigned>
+KOKKOS_FUNCTION constexpr unsigned int to_chars_len(Unsigned val) {
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_StringManipulation.hpp:165]: I was going for reduced number of instantiations (only `unsigned int` or `unsigned long long`)

File: core/unit_test/TestStringManipulation.cpp (hunk 3 above)
Comment (dalg24) [reply core/unit_test/TestStringManipulation.cpp:51]: I opted for undefined behavior

File: core/src/impl/Kokkos_StringManipulation.hpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+#define KOKKOS_STRING_MANIPULATION_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <cstddef>
+#include <type_traits>
+
+namespace Kokkos {
+namespace Impl {
+
+// This header provides a subset of the functionality from <cstring>.  In
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_StringManipulation.hpp:55]: I changed the implementation and do not attempt any more to provide defined behavior when any of the char pointer arguments is null

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_StringManipulation.hpp:55]: For the record I tried to add `KOKKOS_ASSERT` but it does not compile.
```
error: call to non-constexpr function 'void Kokkos::abort(const char*)
```
We print to `std::cerr` in `host_abort` so it is just not going to happen.

**Comment by dalg24** [comment]
Retest this please

**Comment by masterleinad** [comment]
Retest this please.

**Comment by dalg24** [comment]
There are also a lot of `core/src/impl/Kokkos_Error.hpp(117): warning #1305-D: function declared with "noreturn" does return`

**Comment by PhilMiller** [comment]
> There are also a lot of `core/src/impl/Kokkos_Error.hpp(117): warning #1305-D: function declared with "noreturn" does return`

I looked into that a little bit, and I think nvcc is being screwy. The preprocessed source doesn't declare `device_abort` as noreturn

File: core/src/impl/Kokkos_StringManipulation.hpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+// contrast to the standard library header, functions are usable on the device
+// and in constant expressions.  It also includes functionality from <charconv>
+// to convert an integer value to a character sequence.
+
+//<editor-fold desc="String examination">
+// returns the length of a given string
+KOKKOS_FUNCTION constexpr std::size_t strlen(const char *str) {
```
Comment (dalg24) [inline core/src/impl/Kokkos_StringManipulation.hpp:62]: These need to be declared `inline`.  Will fix.
